src/backend/distributed/utils/resource_lock.c

```diff
@@ -958,38 +959,63 @@ LockModeToLockModeText(LOCKMODE lockMode)
 
 
 /*
- * lock_relation_if_exists gets a relation name and lock mode
+ * LockRelationIfExists gets a relation name, lock mode, a nowait flag
  * and returns true if the relation exists and can be locked with
  * the given lock mode. If the relation doesn't exists, the function
  * return false.
  *
+ * If nowait is true, and the lock is already acquired an error will be
+ * thrown.
+ *
  * The relation name should be qualified with the schema name.
  *
  * The function errors out if the lockmode isn't defined in the PostgreSQL's
  * explicit locking table.
  */
-Datum
-lock_relation_if_exists(PG_FUNCTION_ARGS)
+static bool
+LockRelationIfExists(text *relationName, const char *lockModeCString, bool nowait)
```

Comment:
(e.g., at least can you re-run the regression tests with throwing an error if the relation does not exist? If there is something obvious, we could see. )

Reply:
> is it really good to skip if the relation does not exists for the LOCK command? I don't quite remember why it is the case for TRUNCATE as well. Do you know the reason?

we may have had some logic where we would always do lock_relation_if_exists even for local tables, such that we would take this lock before doing anything else, but not quite sure

Reply:
actually, seems like we always only locked distributed tables. Wonder if it's handling a race condition.

Reply:
Ow, we seem to blindly send to worker nodes without checking hasmetadata (not sure why)

Reply:
aah, we did not sync hasmetadata before!

Reply:
Given that we now sync hasmetadata and we expect the table to exist everywhere (or could at least add a hasMetadata check), does this mean we can simply send LOCK instead of changing lock_relation_if_exists?

Reply:
I think that makes a lot of sense, also makes `LOCK VIEW` implementation much simpler. Because, with that, we'd simply send the command instead of calculating which tables to lock (and also avoid acquiring locks per table per roundtrip)

Reply:
> Alternatively, we can construct a command
	SET citus.enable_ddl_propagation TO 'off';
	LOCK ...;
	SET citus.enable_ddl_propagation TO 'off';


Yes, we do this all over the code. And, we should do that here as well. See several usages of
```C

	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
								(void *) commandXXX,
								ENABLE_DDL_PROPAGATION);
```


>  I can do this in a separate PR since this will probably require some more refactoring.

It should not be a large refactor. I think avoiding the view calculation code is much more desirable over the refactor. Besides, it is always good to send the exact same command that user types in `e.g., LOCK ...` instead of Citus generated (e.g., `lock_relation_if_exists`) as there might be subtle minor differences, and relying on PG is much better.

I think we can try the refactor against master branch, to see how large is that. Later, we can use that logic here.